### DISCUSSION TITLE
implementation of a bounded damage model with mfront +castem

### DISCUSSION THREAD

**Discussion by clempham** [thelfer/tfel#628] [Q&A] [answered]
Hello, I would like to implement a bounded gradient damage model for which I have a certain expression of the behavior law in the following form (photo)
![Capture](https://github.com/user-attachments/assets/cf031cd6-590c-418e-a116-3465f56b6669)
Furthermore, I have two stiffness matrices (for which I have the theoretical expressions): one related to displacements and the other to the damage variable. I had started looking into how to do this solely in Cast3M, but I found myself stuck. After looking at some of the examples provided, I thought it would be a good idea to try it in MFront. Could you please guide me? Because I don’t really know where to start...

For now, my script looks like this, But I am lost in the big picture and how, for example, to retrieve the positive and negative parts, etc...

@Description{
  "modele a gradient d'endommagement borné"
}

@MaterialProperty stress young;
young.setGlossaryName("YoungModulus");
@MaterialProperty real nu;
nu.setGlossaryName("PoissonRatio");

@StateVariable real d;
d.setGlossaryName("Damage");
@StateVariable real Y;
Y.setEntryName("DamageDrivingForce");

@LocalVariable stress lambda;
@LocalVariable stress mu;
@LocalVariable real d0;
@LocalVariable StrainStensor e;
@LocalVariable StressStensor s0;

@InitLocalVariables {
  lambda = computeLambda(young, nu);
  mu = computeMu(young, nu);
}

@PredictionOperator{
  if(smt==ELASTIC){
    Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
  } else if(smt==SECANTOPERATOR){
    Dt = (1-d)*(lambda*Stensor4::IxI()+2*mu*Stensor4::Id());
  } else {
    return FAILURE;
  }
}

@ProvidesTangentOperator;

@Integrator{
  //simplifciation opérateur identité
  constexpr const auto id = Stensor::Id();

  // simplifciation operateur trace 
   const strain tr = trace(e);

  // deformation totale
  e = eto + deto ; 

  //valeur propre de la deformation
  e.computeEigenValues(e1,e2,e3);

 // calcul des parties positives et négatives de la déformation ?? A REVOIR
  StrainStensor e_plus = e; // Déformation positive
  StrainStensor e_minus = e; // Déformation négative

 // Calcul des contraintes
  s0 = (1 - d) * (lambda * tr * id + 2 * mu * e_plus + tr * id + 2 * mu * e_minus);

 // Mise à jour de la contrainte dans l'état actuel
  sigma = s0;

 // Mécanisme d'endommagement

  if (Y > threshold) { // threshold est une valeur seuil à définir
    d += rate * (Y - threshold) // à modifier pour évoluer l'endommagement
  }
  // Actualisation de la force d'endommagement
  Y = calculateDamageDrivingForce(e, d); // Fonction à définir selon ton modèle

I am available by email at clemence.pham@cea.fr for more informations about the model

Thank you a lot

**Comment by thelfer** [comment]
Dear @clempham,
You may want to have a look here: https://thelfer.github.io/mgis/web/mgis_fenics_phase_field.html
P.S. To be honest, I did not understand where is the gradient part in your model. 
Best,
Thomas

**Reply by clempham** [reply]
I am currently drawing inspiration from the phase field approach, notably from the phase field damage example 'MFront behaviour for the damage problem', but I get this error message when I try to compile, even though I haven't modified anything... What could this be related to?
![image](https://github.com/user-attachments/assets/c6ed494f-9e68-4ca6-8892-ab262ade8728)

Thank you, 

Clémence

**Reply by thelfer** [reply, marked as answer]
This file is not supported by the `Cast3M` interface. This solver supports material properties, mechanical behaviours (small and finite strain), cohesive zone models and simple point-wise models, but it does not support generalized behaviours.
